grid_map_core/src/Polygon.cpp

```diff
@@ -83,4 +87,98 @@ void Polygon::resetTimestamp()
   timestamp_ = 0.0;
 }
 
+Polygon Polygon::convexHullCircles(const Position center1,
+                                   const Position center2, const double radius,
+                                   const int nVertices)
+{
+  Eigen::Vector2d centerToVertex, centerToVertexTemp;
+  centerToVertex = center2 - center1;
+  centerToVertex.normalize();
+  centerToVertex *= radius;
+
+  grid_map::Polygon polygon;
+  for (int j = 0; j < ceil(nVertices / 2.0); j++) {
+    double theta = M_PI_2 + j * M_PI / (ceil(nVertices / 2.0) - 1);
+    Eigen::Rotation2D<double> rot2d(theta);
+    centerToVertexTemp = rot2d.toRotationMatrix() * centerToVertex;
+    polygon.addVertex(center1 + centerToVertexTemp);
+  }
+  for (int j = 0; j < ceil(nVertices / 2.0); j++) {
+    double theta = 3 * M_PI_2 + j * M_PI / (ceil(nVertices / 2.0) - 1);
+    Eigen::Rotation2D<double> rot2d(theta);
+    centerToVertexTemp = rot2d.toRotationMatrix() * centerToVertex;
+    polygon.addVertex(center2 + centerToVertexTemp);
+  }
+  return polygon;
+}
+
+Polygon Polygon::convexHullCircle(const Position center, const double radius,
+                                  const int nVertices)
+{
+  Eigen::Vector2d centerToVertex(radius, 0.0), centerToVertexTemp;
+
+  grid_map::Polygon polygon;
+  for (int j = 0; j < nVertices; j++) {
+    double theta = j * 2 * M_PI / (nVertices - 1);
+    Eigen::Rotation2D<double> rot2d(theta);
+    centerToVertexTemp = rot2d.toRotationMatrix() * centerToVertex;
+    polygon.addVertex(center + centerToVertexTemp);
+  }
+  return polygon;
+}
+
+Polygon Polygon::convexHull(Polygon& polygon1, Polygon& polygon2)
+{
+  std::vector<Position> vertices1 = polygon1.getVertices();
+  std::vector<Position> vertices2 = polygon2.getVertices();
+  std::vector<Position> vertices;
+  vertices.reserve(vertices1.size() + vertices2.size());
+  vertices.insert(vertices.end(), vertices1.begin(), vertices1.end());
+  vertices.insert(vertices.end(), vertices2.begin(), vertices2.end());
+
+  std::vector<Position> hull(vertices.size());
+
+  // Sort points lexicographically
+  std::sort(vertices.begin(), vertices.end(), sortVertices);
+
+  int k = 0;
+  // Build lower hull
+  for (int i = 0; i < vertices.size(); ++i) {
+    while (k >= 2
+        && computeCrossProduct2D(hull.at(k - 1) - hull.at(k - 2),
+                                 vertices.at(i) - hull.at(k - 2)) <= 0)
+      k--;
+    hull.at(k++) = vertices.at(i);
+  }
+
+  // Build upper hull
+  for (int i = vertices.size() - 2, t = k + 1; i >= 0; i--) {
+    while (k >= t
+        && computeCrossProduct2D(hull.at(k - 1) - hull.at(k - 2),
+                                 vertices.at(i) - hull.at(k - 2)) <= 0)
+      k--;
+    hull.at(k++) = vertices.at(i);
+  }
+  hull.resize(k - 1);
+
+  Polygon polygonOut;
```

Comment:
Use this constructor directly without copying all points individually:

  ``Polygon(std::vector<Position> vertices)``